Sources/BraveWallet/Crypto/Stores/NetworkStore.swift

```diff
@@ -57,61 +73,81 @@ public class NetworkStore: ObservableObject {
-      let chain = await rpcService.network(selectedCoin, origin: nil)
-      // since we are fetch network from JsonRpcService,
-      // we don't need to call `setNetwork` on JsonRpcService
-      self.selectedChainId = chain.chainId
-      // update `isSwapSupported` for Buy/Send/Swap panel
-      self.isSwapSupported = await swapService.isSwapSupported(chain.chainId)
-    }
+  }
+  
+  @MainActor func setup() async {
+    await updateChainList()
+    await updateSelectedChain()
+  }
+  
+  /// Updates the `selectedChainId` and `isSwapSupported` for the network for the current `origin`.
+  @MainActor private func updateSelectedChain() async {
+    // fetch current selected network
+    let selectedCoin = await walletService.selectedCoin()
+    let chain = await rpcService.network(selectedCoin, origin: nil)
+    // since we are fetch network from JsonRpcService,
+    // we don't need to call `setNetwork` on JsonRpcService
+    self.defaultSelectedChainId = chain.chainId
+    let chainForOrigin = await rpcService.network(selectedCoin, origin: origin)
+    self.selectedChainIdForOrigin = chainForOrigin.chainId
+    // update `isSwapSupported` for Buy/Send/Swap panel
+    self.isSwapSupported = await swapService.isSwapSupported(chain.chainId)
   }
 
-  private func updateChainList() {
-    Task { @MainActor in // fetch all networks for all coin types
-      self.allChains = await rpcService.allNetworksForSupportedCoins()
-      
-      let customChainIds = await rpcService.customNetworks(.eth) // only support Ethereum custom chains
-      self.customChains = allChains.filter { customChainIds.contains($0.id) }
-    }
+  @MainActor private func updateChainList() async {
+    // fetch all networks for all coin types
+    self.allChains = await rpcService.allNetworksForSupportedCoins()
+    
+    let customChainIds = await rpcService.customNetworks(.eth) // only support Ethereum custom chains
+    self.customChains = allChains.filter { customChainIds.contains($0.id) }
   }
   
   func isCustomChain(_ network: BraveWallet.NetworkInfo) -> Bool {
     customChains.contains(where: { $0.coin == network.coin && $0.chainId == network.chainId })
   }
   
-  @MainActor @discardableResult func setSelectedChain(_ network: BraveWallet.NetworkInfo) async -> SetSelectedChainError? {
+  @MainActor @discardableResult func setSelectedChain(
+    _ network: BraveWallet.NetworkInfo,
+    isForOrigin: Bool
+  ) async -> SetSelectedChainError? {
     let keyringId = network.coin.keyringId
     let keyringInfo = await keyringService.keyringInfo(keyringId)
     if keyringInfo.accountInfos.isEmpty {
       // Need to prompt user to create new account via alert
       return .selectedChainHasNoAccounts
     } else {
       let selectedCoin = await walletService.selectedCoin()
-      if self.selectedChainId != network.chainId {
-        self.selectedChainId = network.chainId
+      if isForOrigin {
+        if self.selectedChainIdForOrigin != network.chainId {
```

Comment:
if isForOrigin && self.selectedChainIdForOrigin != network.chainId {
} else if self.defaultSelectedChainId != network.chainId {
}

Reply:
Updated in [084d05c](https://github.com/brave/brave-ios/pull/7470/commits/084d05c615b0b0bf277b3bb1132d0202a73febcc)